### PR TITLE
stub: add support for dtb addons

### PR DESCRIPTION
**Pull request by bluca**

### CODE REVIEW

File: src/boot/efi/stub.c [hunk 1]
```diff
@@ -632,13 +697,34 @@ static EFI_STATUS run(EFI_HANDLE image) {
...
+
+        for (size_t i = 0; i < n_dts_addons_uki; ++i) {
+                err = devicetree_install_from_memory(
+                                &dt_state,
+                                dt_bases_addons_uki[i],
+                                dt_sizes_addons_uki[i]);
+                if (err != EFI_SUCCESS)
+                        log_error_status(err, "Error loading local addon devicetree: %m");
+                dt_bases_addons_uki[i] = mfree(dt_bases_addons_uki[i]);
+        }
```
Comment (poettering) [inline src/boot/efi/stub.c]: maybe add a helper: install_dt_addons() that takes the array and the number of entries as parameters. i.e. unify these two almost identical for loops in a single function

File: src/boot/efi/stub.c [hunk 2]
```diff
@@ -248,25 +248,52 @@ static EFI_STATUS load_addons_from_dir(
 
 }
 
-static EFI_STATUS cmdline_append_and_measure_addons(
+static void cmdline_append_and_measure_addons(
+                char16_t *cmdline,
+                bool *ret_parameters_measured,
+                char16_t **cmdline_append) {
```
Comment (poettering) [inline src/boot/efi/stub.c]: maybe swap the two args. return params are usuall last.

File: src/boot/efi/stub.c [hunk 3]
```diff
@@ -345,36 +377,59 @@ static EFI_STATUS cmdline_append_and_measure_addons(
...
+                *ret_dt_bases = TAKE_PTR(dt_bases);
+                *ret_dt_sizes = TAKE_PTR(dt_sizes);
+                *ret_n_dt = n_dt;
         }
 
         return EFI_SUCCESS;
+
+fail:
+        for (size_t i = 0; i < n_dt; ++i)
+                free(dt_bases[i]);
```
Comment (poettering) [inline src/boot/efi/stub.c]: hmm, use CLEANUP_ARRAY()? (and add it to uefi headers if missing there?)

File: src/boot/efi/stub.c [hunk 4]
```diff
@@ -345,36 +377,59 @@ static EFI_STATUS cmdline_append_and_measure_addons(
...
+                if (ret_dt_bases && szs[UNIFIED_SECTION_DTB] > 0) {
+                        dt_sizes = xrealloc(dt_sizes,
+                                            n_dt * sizeof(size_t),
+                                            (n_dt + 1)  * sizeof(size_t));
+                        dt_sizes[n_dt] = szs[UNIFIED_SECTION_DTB];
+
+                        dt_bases = xrealloc(dt_bases,
+                                            n_dt * sizeof(EFI_PHYSICAL_ADDRESS),
+                                            (n_dt + 1) * sizeof(EFI_PHYSICAL_ADDRESS));
+                        dt_bases[n_dt] = xmalloc(dt_sizes[n_dt]);
```
Comment (poettering) [inline src/boot/efi/stub.c]: please add xmemdup()

File: src/boot/efi/stub.c (hunk 3 above)
Comment (bluca) [reply src/boot/efi/stub.c]: I tried to move it but I'm getting an error I don't undestand:

```
../src/boot/efi/stub.c:344:9: error: cleanup argument not a function
  344 |         CLEANUP_ARRAY(dt_bases, n_dt, dt_array_free);
```

I have no idea why it says it's not a function, when it clearly is, with prototype `static void dt_array_free(void **dt_bases, size_t n_dts) `

**Comment by bluca** [comment]
I can't fully test this as whatever I do I'm getting a kernel panic (without this PR) when booting u-boot on qemu-arm64 (which is the only thing that supports the device tree UEFI protocol). But debug prints show that the protocol loading is successful for both embedded and addon DTBs, so I think that's the best I can do

File: man/systemd-stub.xml [hunk 5]
```diff
@@ -161,19 +161,20 @@
...
-      filename, so that, given the same set of addons, the same set of measurements can be expected in
-      PCR12, however note that the filename is not protected by the PE signature, and as such an attacker
-      with write access to the ESP could potentially rename these files to change the order in which they
-      are loaded, in a way that could alter the functionality of the kernel, as some options might be order
-      dependent. If you sign such addons, you should pay attention to the PCR12 values and make use of an
-      attestation service so that improper use of your signed addons can be detected and dealt with using
-      one of the aforementioned revocation mechanisms.</para></listitem>
+      Addons are supposed to be used to pass additional kernel command line parameters or devite tree blobs,
+      regardless of the kernel image being booted, for example to allow platform vendors to ship
+      platform-specific configuration. The loaded command line addon files are sorted, loaded, measured into
```
Comment (keszybz) [inline man/systemd-stub.xml]: "and measured"

Edit: the first three comments are about the existing text. If they are addressed, this should be made into a separate commit.

File: man/systemd-stub.xml [hunk 6]
```diff
@@ -161,19 +161,20 @@
...
+      TPM PCR 12 (if a TPM is present) and appended to the kernel command line. UKI command line options are
+      listed first, then options from addons in <filename>/loader/addons/*.addon.efi</filename> are appended
+      next, and finally UKI-specific addons are appended last. Device tree blobs are loaded following the
```
Comment (keszybz) [inline man/systemd-stub.xml]: Strike "are appended next" and "are appended last".

File: man/systemd-stub.xml [hunk 7]
```diff
@@ -161,19 +161,20 @@
...
+      same algorithm, but they are not measured by the stub. Addons are always loaded in the same order
+      based on the filename, so that, given the same set of addons, the same set of measurements can be
+      expected in PCR12, however note that the filename is not protected by the PE signature, and as such an
```
Comment (keszybz) [inline man/systemd-stub.xml]: It'd be nicer if "however" started a new sentence here.

File: man/systemd-stub.xml (hunk 5 above)
Comment (bluca) [reply man/systemd-stub.xml]: why a separate commit?

File: src/boot/efi/stub.c (hunk 3 above)
Comment (poettering) [reply src/boot/efi/stub.c]: did you figure that out ever?

File: src/boot/efi/stub.c [hunk 8]
```diff
@@ -249,25 +249,71 @@ static EFI_STATUS load_addons_from_dir(
...
+
+        EFI_STATUS err;
+
+        assert(dt_state);
+        assert(n_dts == 0 || (dt_bases && dt_sizes));
+
+        for (size_t i = 0; i < n_dts; ++i) {
+                err = devicetree_install_from_memory(dt_state, dt_bases[i], dt_sizes[i]);
+                if (err != EFI_SUCCESS)
+                        log_error_status(err, "Error loading global addon devicetree: %m");
```
Comment (poettering) [inline src/boot/efi/stub.c]: ", ignoring"

File: src/boot/efi/stub.c (hunk 3 above)
Comment (bluca) [reply src/boot/efi/stub.c]: No I couldn't make it work, after bashing it with a hammer for an hour or so I gave up. If anybody knows how to make it work I'm happy to take the change, but have no idea how otherwise

File: src/boot/efi/stub.c [hunk 9]
```diff
@@ -467,27 +567,10 @@ static EFI_STATUS run(EFI_HANDLE image) {
...
-        err = cmdline_append_and_measure_addons(
-                        image,
-                        loaded_image,
-                        dropin_dir,
-                        uname,
-                        &m,
-                        &cmdline);
-        if (err != EFI_SUCCESS)
-                log_error_status(err, "Error loading UKI-specific addons, ignoring: %m");
+        cmdline_append_and_measure_addons(cmdline_addons_uki, &cmdline, &m);
```
Comment (poettering) [inline src/boot/efi/stub.c]: hmm, actually it kinda sucks generating multiple measurement records for these measurements. I'd prefer if we could merge that into one measurement. I udnerstand that this is a pre-existing problem, but maybe fix that when you touch this

It's particularly bad that we can't distinguish the measurements, i.e.we measure the global and the per-uki cmdlines separately, but we cannot determine afterwards if we have one measurement which of the two it was. Hence we have the distinction that is useless there.

I think we should improve this at least a tiny bit here, and merge the per-uki and global cmdline measurement into one, i.e. merge *all* add-on cmdline additions first, measure them, and then append them to the overall cmdline.

This kinda matters for me for the pcrlock work: i must have a sane way to reproduce the measurements...

(I mean, it's still going to be chaotic enough, as we will have three undiscernable measurements in the worst case: addon cmdline, smbios cmdline and process cmdline).

File: src/boot/efi/stub.c (hunk 3 above)
Comment (poettering) [reply src/boot/efi/stub.c]: did you actually copy array_cleanup() static inline func over into fundamental/ too?

Comment (poettering) [reply src/boot/efi/stub.c]: (i mean, `array_cleanup()` is the function used as cleanup argument, not the one you pass

**Comment by poettering** [comment]
Hmm, so I am pretty sure the dtbs we load from addons should be measured as such, probably into PCR 12.

File: src/boot/efi/stub.c (hunk 3 above)
Comment (bluca) [reply src/boot/efi/stub.c]: yes, I copied macro and static helper, and when compiling for !efi it did not complain at all, it was extremely weird

Comment (poettering) [reply src/boot/efi/stub.c]: Seems to just work here: 

https://paste.centos.org/view/raw/dbff36e8

Comment (poettering) [reply src/boot/efi/stub.c]: (tested that with both gcc and clang, works perfectly on both)

Comment (bluca) [reply src/boot/efi/stub.c]: interesting - I no longer have saved what I was doing so no idea if it is different and how, I'll apply that when I get to it and try again, thanks

**Comment by poettering** [comment]
btw, i wonder what the relationship to #28959.

i.e. if we have devicetree addons, should we filter them by compatibility too?

File: src/boot/efi/stub.c (hunk 9 above)
Comment (bluca) [reply src/boot/efi/stub.c]: ok, done

**Comment by bluca** [comment]
> btw, i wonder what the relationship to #28959.
> 
> i.e. if we have devicetree addons, should we filter them by compatibility too?

let's leave that for later, when that pr is merged

File: src/boot/efi/stub.c [hunk 10]
```diff
@@ -350,22 +445,37 @@ static EFI_STATUS cmdline_append_and_measure_addons(
...
 
-        if (!isempty(buffer)) {
-                _cleanup_free_ char16_t *tmp = TAKE_PTR(*cmdline_append);
-                bool m = false;
+                        dt_bases = xrealloc(dt_bases,
+                                            n_dt * sizeof(EFI_PHYSICAL_ADDRESS),
+                                            (n_dt + 1) * sizeof(EFI_PHYSICAL_ADDRESS));
+                        dt_bases[n_dt] = xmemdup(
+                                PHYSICAL_ADDRESS_TO_POINTER(POINTER_TO_PHYSICAL_ADDRESS(loaded_addon->ImageBase) + addrs[UNIFIED_SECTION_DTB]),
+                                dt_sizes[n_dt]);
```
Comment (medhefgo) [inline src/boot/efi/stub.c]: This should use `void*` for calculations as it's the type of `dt_bases`.

Comment (bluca) [reply src/boot/efi/stub.c]: What do you mean precisely?

Comment (medhefgo) [reply src/boot/efi/stub.c]: The type of `dt_bases` is `void*`, so why go through `EFI_PHYSICAL_ADDRESS`?
```suggestion
                        dt_bases = xrealloc(dt_bases,
                                            n_dt * sizeof(void*),
                                            (n_dt + 1) * sizeof(void*));
                        dt_bases[n_dt] = xmemdup((uint8_t*)loaded_addon->ImageBase + addrs[UNIFIED_SECTION_DTB], dt_sizes[n_dt]);
```

Comment (bluca) [reply src/boot/efi/stub.c]: ah in that sense - no reason, fixed

Comment (medhefgo) [reply src/boot/efi/stub.c]: You missed one `sizeof(EFI_PHYSICAL_ADDRESS)`.

Comment (bluca) [reply src/boot/efi/stub.c]: whops

File: src/boot/efi/util.h [hunk 11]
```diff
@@ -52,6 +52,12 @@ static inline void *xrealloc(void *p, size_t old_size, size_t new_size) {
         return t;
 }
 
+_malloc_ _alloc_(2) _returns_nonnull_ _warn_unused_result_
+static inline void* xmemdup(const void *p, size_t l) {
+        void *t = xmalloc(l);
+        return memcpy(t, p, l);
```
Comment (poettering) [inline src/boot/efi/util.h]: you could shorten this to:

```c
return memcpy(xmalloc(l), p, l);
```

File: man/systemd-stub.xml [hunk 12]
```diff
@@ -167,19 +167,20 @@
...
+      are listed first, then options from addons in <filename>/loader/addons/*.addon.efi</filename>, and
+      finally UKI-specific addons. Device tree blobs are loaded and measured following the same algorithm.
+      Addons are always loaded in the same order based on the filename, so that, given the same set of
+      addons, the same set of measurements can be expected in PCR12. However, note that the filename is not
+      protected by the PE signature, and as such an attacker with write access to the ESP could potentially
+      rename these files to change the order in which they are loaded, in a way that could alter the
+      functionality of the kernel, as some options might be order dependent. If you sign such addons, you
+      should pay attention to the PCR12 values and make use of an attestation service so that improper use
+      of your signed addons can be detected and dealt with using one of the aforementioned revocation
+      mechanisms.</para></listitem>
```
Comment (poettering) [inline man/systemd-stub.xml:183]: i would feel a ton more comfortable if we'd just include a list of permitted file names in the PE file somewhere, and then refuse PE files where this does not match. Seems awfully handwavy to put the work on the users

Comment (bluca) [reply man/systemd-stub.xml:183]: that is a more generic thing and not really related to DTBs, let's leave that for another PR please

File: man/systemd-stub.xml [hunk 13]
```diff
@@ -188,9 +189,9 @@
       measured into TPM PCR 12 (if a TPM is present).</para></listitem>
 
       <listitem><para>Additionally, files <filename>/loader/addons/*.addon.efi</filename> are loaded and
-      verified as PE binaries, and a <literal>.cmdline</literal> section is parsed from them. This is
-      supposed to be used to pass additional command line parameters to the kernel, regardless of the kernel
-      being booted.</para></listitem>
+      verified as PE binaries, and <literal>.cmdline</literal> and/or <literal>.dtb</literal> sections are
+      parsed from them. This is supposed to be used to pass additional command line parameters or device
+      tree blobs to the kernel, regardless of the kernel being booted.</para></listitem>
```
Comment (poettering) [inline man/systemd-stub.xml]: the dt spec spells this as "Devicetree" btw, not "device tree" or "Device Tree" or "DeviceTree" or so.

File: src/boot/efi/stub.c [hunk 14]
```diff
@@ -253,29 +254,113 @@ static EFI_STATUS load_addons_from_dir(
...
+static void cmdline_append_and_measure_addons(
+                char16_t *cmdline,
+                char16_t **cmdline_append,
+                bool *ret_parameters_measured) {
+
+        _cleanup_free_ char16_t *tmp = NULL;
+        bool m = false;
+
+        assert(ret_parameters_measured);
+        assert(cmdline_append);
```
Comment (poettering) [inline src/boot/efi/stub.c:268]: doesn't matter, but we usually add the asserts for func params in the order the function takes the parameters. i.e. swap these two lines?

File: src/boot/efi/stub.c [hunk 15]
```diff
@@ -253,29 +254,113 @@ static EFI_STATUS load_addons_from_dir(
 
 }
```
Comment (poettering) [inline src/boot/efi/stub.c:258]: btw, please add a new EFI_STUB_FEATURE_XYZ feature flag for this, and display it in bootctl.

i.e. EFI_STUB_FEATURE_DEVICETREE_ADDONS or so, mirroring the existing EFI_STUB_FEATURE_CMDLINE_ADDONS

File: src/boot/efi/stub.c [hunk 16]
```diff
@@ -253,29 +254,113 @@ static EFI_STATUS load_addons_from_dir(
...
+        EFI_STATUS err;
+
+        assert(dt_state);
+        assert(n_dts == 0 || (dt_bases && dt_sizes));
+        assert(ret_parameters_measured);
+
+        for (size_t i = 0; i < n_dts; ++i) {
+                err = devicetree_install_from_memory(dt_state, dt_bases[i], dt_sizes[i]);
+                if (err != EFI_SUCCESS)
+                        log_error_status(err, "Error loading global addon devicetree, ignoring: %m");
```
Comment (poettering) [inline src/boot/efi/stub.c]: "global" is not necessarily correct.

File: src/boot/efi/stub.c [hunk 17]
```diff
@@ -253,29 +254,113 @@ static EFI_STATUS load_addons_from_dir(
...
+                else {
+                        _cleanup_free_ char16_t* description = NULL;
+                        bool m = false;
+
+                        description = xasprintf("DTB addon #%zu", i);
```
Comment (poettering) [inline src/boot/efi/stub.c]: please don't do this, this is useless information.

Comment (bluca) [reply src/boot/efi/stub.c]: ok, what should I use for 'description' here?

File: src/boot/efi/stub.c [hunk 18]
```diff
@@ -253,29 +254,113 @@ static EFI_STATUS load_addons_from_dir(
...
+
+                        err = tpm_log_event(
+                                        TPM2_PCR_KERNEL_CONFIG,
+                                        POINTER_TO_PHYSICAL_ADDRESS(dt_bases[i]),
+                                        dt_sizes[i],
+                                        description,
+                                        &m);
```
Comment (poettering) [inline src/boot/efi/stub.c:319]: so the way we so far did TPM event logs is terrible, because it makes is very hard to recognize stuff in the event logs, and that's important for systemd-pcrlock and similar tools.

Given you are measuring something we haven't measured before, let's please fix this, and measure things in a more useful, recognizable way.

Specifically this means tpm_log_event() should be improved to optionally not measure things as an EV_IPL event, but a "tagged event", i.e. EV_EVENT_TAG, with a recognizable tag. See how the kernel does this via INITRD_EVENT_TAG_ID + LOAD_OPTIONS_EVENT_TAG_ID.

And then, in then in the event data say the file name of what you are measuring here.

File: src/boot/efi/stub.c (hunk 17 above)
Comment (poettering) [reply src/boot/efi/stub.c]: filename, see below.

File: src/boot/efi/stub.c (hunk 18 above)
Comment (poettering) [reply src/boot/efi/stub.c:319]: instead of improving tpm_log_event() maybe add an alternative call tpm_log_event_tagged() or so

Comment (poettering) [reply src/boot/efi/stub.c:319]: (and add a comment in tpm_log_event() it should not be used for new measurements)

Comment (bluca) [reply src/boot/efi/stub.c:319]: I'll use the filename for the description. While I agree that we should fix how we do event logs, testing this PR is already difficult and tedious as it is - only partially works on arm64 which is slow as hell to build plus uboot which causes a kernel panic, so the only way to have some bare idea whether things are somewhat working is with debug prints. So I am not going to refactor the event logs stuff here, as it's a generic problem that affects the rest of the stub/loader too, and we can do it separately and on a more sane and usable environment

Comment (poettering) [reply src/boot/efi/stub.c:319]: sure, this can happen in preparatory step. But note that there's plenty of code that uses our existing measurements already, hence for all the stuff we already measure we cannot really change the records we generate. Which is very unfortunate, but people have learnt to live with it.

But for the future we should do better, hence all new measurements shoudl use EV_EVENT_TAG event log records

Comment (medhefgo) [reply src/boot/efi/stub.c:319]: > only partially works on arm64 which is slow as hell to build

Are you building in a VM/qemu-user chroot? I've recently learned how to do debian multiarch. Then you can easily cross-build with native speed.

> so the only way to have some bare idea whether things are somewhat working is with debug prints

Debugging sd-boot is possible: tools/debug-sd-boot.sh. Also, you may wanna build with `-Db_sanitize=undefined` to make sure there isn't any UB involved on our side.

File: man/systemd-stub.xml [hunk 19]
```diff
@@ -167,19 +167,20 @@
...
-      then options from addons in <filename>/loader/addons/*.addon.efi</filename> are appended next, and
-      finally UKI-specific addons are appended last. Addons are always loaded in the same order based on the
-      filename, so that, given the same set of addons, the same set of measurements can be expected in
-      PCR12, however note that the filename is not protected by the PE signature, and as such an attacker
-      with write access to the ESP could potentially rename these files to change the order in which they
-      are loaded, in a way that could alter the functionality of the kernel, as some options might be order
-      dependent. If you sign such addons, you should pay attention to the PCR12 values and make use of an
-      attestation service so that improper use of your signed addons can be detected and dealt with using
-      one of the aforementioned revocation mechanisms.</para></listitem>
+      Addons are supposed to be used to pass additional kernel command line parameters or devite tree blobs,
```
Comment (RaitoBezarius) [inline man/systemd-stub.xml]:
```suggestion
      Addons are supposed to be used to pass additional kernel command line parameters or device tree blobs,
```

**Comment by poettering** [comment]
the event log records this adds are still EV_IPL, please address this.

**Comment by poettering** [comment]
(i don't think this should be merged with bad tpm event logging, sorry)

**Comment by bluca** [comment]
> the event log records this adds are still EV_IPL, please address this.

As already mentioned, I am not able to get past a kernel panic with a firmware that supports DTB, so I cannot possibly test changing TPM logging format. So either you are fine with merging a completely untested refactor of TPM logging, or it stays as-is.

**Comment by bluca** [comment]
> > the event log records this adds are still EV_IPL, please address this.
> 
> As already mentioned, I am not able to get past a kernel panic with a firmware that supports DTB, so I cannot possibly test changing TPM logging format. So either you are fine with merging a completely untested refactor of TPM logging, or it stays as-is.

Ok, let's go with 100% untested then. No idea if it works or how, as it's not possible to test it without hardware.

File: src/boot/efi/proto/tcg.h [hunk 20]
```diff
@@ -11,6 +11,7 @@
 #define TCG_ALG_SHA 0x4
 #define EFI_TCG2_EVENT_HEADER_VERSION 1
 #define EV_IPL 13
+#define EV_EVENT_TAG 0x6
```
Comment (poettering) [inline src/boot/efi/proto/tcg.h]: it's a bit weird that even through EV_IPL and EV_EVENT_TAG are values for the same field one is encoded in hex, and one in dec. please convert both to hex or both to dec...

also, maybe change this to be unsigned values, given the fields for this are unsigned too

Comment (poettering) [reply src/boot/efi/proto/tcg.h]: UINT32_C() maybe even

File: src/boot/efi/measure.c [hunk 21]
```diff
@@ -42,6 +42,51 @@ static EFI_STATUS tpm1_measure_to_pcr_and_event_log(
...
+
+        assert(tcg);
+        assert(description);
+
+        desc_len = strsize16(description);
+        event_size = offsetof(EFI_TCG2_EVENT, Event) + offsetof(EFI_TCG2_TAGGED_EVENT, Event) + desc_len;
+
+        event = xmalloc(event_size);
+
+        memset(event, 0, event_size);
```
Comment (poettering) [inline src/boot/efi/measure.c]: pleas rebase on https://github.com/systemd/systemd/pull/29405 and use xmalloc0() here

File: src/boot/efi/stub.c [hunk 22]
```diff
@@ -25,6 +26,8 @@ DECLARE_NOALLOC_SECTION(".sdmagic", "#### LoaderInfo: systemd-stub " GIT_VERSION
 
 DECLARE_SBAT(SBAT_STUB_SECTION_TEXT);
 
+#define ADDON_FILENAME_EVENT_TAG_ID 0x6c46f751
```
Comment (poettering) [inline src/boot/efi/stub.c]: please enclose in UINT32_C()

where does the id come from? i hope /dev/urandom?

Comment (bluca) [reply src/boot/efi/stub.c]: openssl rand -hex 16

File: src/boot/efi/measure.c (hunk 21 above)
Comment (poettering) [reply src/boot/efi/measure.c]: #29405 has been merged now

File: src/boot/efi/measure.c [hunk 23]
```diff
@@ -42,6 +42,51 @@ static EFI_STATUS tpm1_measure_to_pcr_and_event_log(
...
+                uint32_t pcrindex,
+                EFI_PHYSICAL_ADDRESS buffer,
+                uint64_t buffer_size,
+                uint32_t event_id,
+                const char16_t *description) {
+
+        _cleanup_free_ struct event {
+                EFI_TCG2_EVENT tcg_event;
+                EFI_TCG2_TAGGED_EVENT tcg_tagged_event;
+        } *event;
```
Comment (medhefgo) [inline src/boot/efi/measure.c]: This should really be NULLed, or better yet, move the definition to the first use. It's not the 80s anymore.

**Comment by poettering** [comment]
Hmm, now that we have it, can you also extend docs/TPM2_PCR_MEASUREMENTS.md with the new measurement.

**Comment by bluca** [comment]
> as mentioned elsewhere, might be good to change config_load_defaults() in boot.c to also call the new tagged measurement function, and measure the boot loader config file we loaded after we loaded it. a one liner addition that makes sure the new measurement func works correctly.
> 
> should measure to TPM2_PCR_BOOT_LOADER_CONFIG (aka PCR 5) which we currently don't measure anything into.
> 
> you could then drop an item form the TODO file, too: `* sd-boot should probably measure its configuration file to PCR 5 at boot, as per TCG PC Client Platform Firmware Profile Spec.`

it seems fragile to measure the whole content of loader.conf, given we ship it with comments for defaults, so if we add something new and ship a commented out default we'll break boot for everyone that uses PCR5... but whatever, I've just added it, I just want to get this done as it's been piling more and more and more stuff in it for 3 months

File: src/boot/efi/measure.c [hunk 24]
```diff
@@ -42,6 +42,50 @@ static EFI_STATUS tpm1_measure_to_pcr_and_event_log(
...
+                uint32_t pcrindex,
+                EFI_PHYSICAL_ADDRESS buffer,
+                uint64_t buffer_size,
+                uint32_t event_id,
+                const char16_t *description) {
+
+        _cleanup_free_ struct event {
+                EFI_TCG2_EVENT tcg_event;
+                EFI_TCG2_TAGGED_EVENT tcg_tagged_event;
+        } *event = NULL;
```
Comment (poettering) [inline src/boot/efi/measure.c]: might be wise to use `_packed_` here

File: src/boot/efi/stub.c [hunk 25]
```diff
@@ -623,13 +632,10 @@ static EFI_STATUS run(EFI_HANDLE image) {
...
         /* If we have any extra command line to add via PE addons, load them now and append, and
-         * measure the additions separately, after the embedded options, but before the smbios ones,
+         * measure the additions together, after the embedded options, but before the smbios ones,
          * so that the order is reversed from "most hardcoded" to "most dynamic". The global addons are
          * loaded first, and the image-specific ones later, for the same reason. */
-        cmdline_append_and_measure_addons(cmdline_addons_global, &cmdline, &m);
-        parameters_measured = parameters_measured < 0 ? m : (parameters_measured && m);
-
-        cmdline_append_and_measure_addons(cmdline_addons_uki, &cmdline, &m);
+        cmdline_append_and_measure_addons(cmdline_addons_global, cmdline_addons_uki, &cmdline, &m);
```
Comment (poettering) [inline src/boot/efi/stub.c:638]: i think this makes a ton of sense, and even though strictly speaking this might be considered a compat break I'd just do it. But please update docs/TPM2_PCR_MEASUREMENTS.md accordingly, which claims we'd do two separate measurements currently
